app/src/main/java/com/appdev/eateryblueandroid/util/LoginRepository.kt

```diff
@@ -0,0 +1,295 @@
+package com.appdev.eateryblueandroid.util
+
+import android.util.Log
+import com.appdev.eateryblueandroid.models.*
+import com.appdev.eateryblueandroid.ui.appContext
+import com.appdev.eateryblueandroid.ui.viewmodels.ProfileViewModel
+import com.appdev.eateryblueandroid.util.Constants.passwordAlias
+import com.appdev.eateryblueandroid.util.Constants.userPreferencesStore
+import com.codelab.android.datastore.AccountProto
+import com.codelab.android.datastore.Date
+import com.codelab.android.datastore.TransactionProto
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+import java.time.LocalDateTime
+
+object LoginRepository {
+    private var loadedUsername: String? = null
+    private var loadedPassword: String? = null
+
+    private var accountsFlow: MutableStateFlow<MutableList<Account>> = MutableStateFlow(mutableListOf())
+    var accounts = accountsFlow.asStateFlow()
+
+    private var transactionsFlow : MutableStateFlow<MutableList<Transaction>> = MutableStateFlow(mutableListOf())
+    var transactions = transactionsFlow.asStateFlow()
+
+    /** True if the caching process has completed fully. False otherwise. */
+    private var cachedFlow : MutableStateFlow<Boolean> = MutableStateFlow(false)
+    var cached = cachedFlow.asStateFlow()
+
+    /**
+     * Attempts to auto login. Only succeeds if loadedUsername AND loadedPassword are non-null
+     * and non-empty, AND any kind of login has not already taken place.
+     */
+    private fun attemptAutoLogin(profileViewModel: ProfileViewModel) {
+        if (profileViewModel.state.value !is ProfileViewModel.State.LoggingIn
+            && profileViewModel.state.value !is ProfileViewModel.State.ProfileData
+            && profileViewModel.state.value !is ProfileViewModel.State.AutoLoggingIn
+            && cached.value && !loadedUsername.isNullOrEmpty() && !loadedPassword.isNullOrEmpty()
+        ) {
+            CoroutineScope(Dispatchers.Default).launch {
+                profileViewModel.autoLogin(
+                    loadedUsername!!,
+                    decryptData(passwordAlias, loadedPassword!!)
+                )
+                Log.i(
+                    "Login",
+                    "Attempting Login with username " + loadedUsername + " and password " + loadedPassword
+                )
+            }
+        }
+    }
+
+    /**
+     * Asynchronously pulls login net ID and password from proto datastore (local storage),
+     * then attempts to log in.
+     *
+     * @see attemptAutoLogin
+     */
+    fun initializeLoginData(profileViewModel: ProfileViewModel) {
```

Comment:
The repositories simply interact with the backend calls and pass the data along to the ViewModels but they shouldn't reference any viewmodels in particular (as the repositories are meant to be able to be used across multiple ViewModels)!

Reply:
See here for a more concrete example. Try and extract out the logic not pertaining to interacting with your data sources and leaving it with the ViewModels themselves

Reply:
Good catch--amended in commit [3021b4e](https://github.com/cuappdev/eatery-blue-android/pull/22/commits/3021b4e54b9c59bb8de783bdf34048f2e8359a48). Let me know what you think of the changes.

Reply:
Looks good !